src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/AddressHeaderCollectionElement.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Configuration;
+using System.Security;
+using System.Xml;
+using CoreWCF.Channels;
+
+namespace CoreWCF.Configuration
+{
+    public sealed class AddressHeaderCollectionElement : ServiceModelConfigurationElement
+    {
+        public AddressHeaderCollectionElement()
+        {
+        }
+
+        internal void Copy(AddressHeaderCollectionElement source)
+        {
+            if (source == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("source");
+            }
+
+            PropertyInformationCollection properties = source.ElementInformation.Properties;
+            if (properties[ConfigurationStrings.Headers].ValueOrigin != PropertyValueOrigin.Default)
+            {
+                this.Headers = source.Headers;
+            }
+        }
+
+        [ConfigurationProperty(ConfigurationStrings.Headers, DefaultValue = null)]
+        public AddressHeaderCollection Headers
+        {
+            get
+            {
+                AddressHeaderCollection retVal = (AddressHeaderCollection)base[ConfigurationStrings.Headers];
+                if (null == retVal)
+                {
+                    retVal = new AddressHeaderCollection();
+                }
+                return retVal;
+            }
+            set
+            {
+                if (value == null)
+                {
+                    value = new AddressHeaderCollection();
+                }
+                base[ConfigurationStrings.Headers] = value;
+            }
+        }
+
+        [SecuritySafeCritical]
+        protected override void DeserializeElement(XmlReader reader, bool serializeCollectionKey)
+        {
+            DeserializeElementCore(reader);
+        }
+
+        private void DeserializeElementCore(XmlReader reader)
+        {
+            //TODO Find proper way to handle this this.Headers = AddressHeaderCollection.ReadServiceParameters(XmlDictionaryReader.CreateDictionaryReader(reader));
+        }
```

Comment:
I think the easiest solution is going to be to make your own type derived from AddressHeader which stores the header. If you look at the .NET Framework implementation, it has an internal class called BufferedAddressHeader. The problem with that is it uses a really efficient XmlBuffer class to hold the header, but that's not public. As this is only config and it will only read once, you can get away with storing it as an XmlElement. I haven't tested this, but I think this should work:
```c#
internal class XmlElementBackedAddressHeader : AddressHeader
{
    private XmlElement _header;

    public XmlElementBackedAddressHeader(XmlDictionaryReader reader)
    {
        Name = reader.LocalName;
        Namespace = reader.NamespaceURI;
        using (XmlReader subTreeReader = reader.ReadSubtree())
        {
            XmlDocument doc = new XmlDocument();
            doc.Load(subTreeReader);
            _header = doc.DocumentElement;
        }
    }
    public override string Name { get; }
    public override string Namespace { get; }

    protected override void OnWriteAddressHeaderContents(XmlDictionaryWriter writer)
    {
        _header.WriteTo(writer);
    }
}
```
Then the code for DeserializeElementCore would be something like:
```c#
Headers = new AddressHeaderCollection(new [] { new XmlElementBackedAddressHeader(reader) });
```

Reply:
Thank you for the direction.  I have implemented this.  I just need to write a test.